src/main.rs

```diff
@@ -1,3 +1,157 @@
+struct Mips {
+    regs: [u32; 32],
+    floats: [f32; 32],
+    mult_hi: u32,
+    mult_lo: u32,
+    pc: usize,
+
+    // Enough for six MIPS instructions.
+    memory: [u32; 13]
+}
+
+struct Rtype {
+    rs: usize,
+    rt: usize,
+    rd: usize,
+    shamt: u8,
+    funct: u8
+}
+
+struct Itype {
+    opcode: u32,
+    rs: usize,
+    rt: usize,
+    imm: u16
+}
+
+// struct Jtype
+// struct Ftype
+
+enum Instructions {
+    R(Rtype),
+    I(Itype),
+    //J and F type
+}
+
+impl Mips {
+
+    fn dispatch_r(&mut self, ins: Rtype) {
+        match ins.funct {
+            // Shift-left logical
+            0x0 => {
+                self.regs[ins.rd] = self.regs[ins.rt] << ins.shamt;
+            }
+            // Shift-right logical
+            0x2 => {
+                self.regs[ins.rd] = self.regs[ins.rt] >> ins.shamt;
+            }
+            // Add
+            0x20 => {
+                self.regs[ins.rd] = self.regs[ins.rt] + self.regs[ins.rs];
+                //Todo- catch overflows
+            }
+            // Subtract
+            0x22 => {
+                //Todo- catch overflows
+                self.regs[ins.rd] = self.regs[ins.rt] + self.regs[ins.rs];
+            }
+            // Xor
+            0x26 => {
+                self.regs[ins.rd] = self.regs[ins.rt] ^ self.regs[ins.rs];
+            }
+            _ => panic!("R-Type unimplemented instruction")
```

Comment:
Didn't we say no panics?

Reply:
we did. I need to plumb errors in everywhere.